pages/JuiceShop/LoginPage.js

```diff
@@ -0,0 +1,43 @@
+const BasePage = require('../basePage.js');
+
+class LoginPage extends BasePage {
+
+    get emailField() {return browser.$$('#login-form input').then(el => el[0]);}
+
+    get passwordField() {return browser.$$('#login-form input').then(el => el[1]);}
+
+    get loginButton() {return browser.$('#login-form #loginButton');}
+
+    get welcomeDialog() {return browser.$('.cdk-overlay-pane');}
+
+    get dismissButton() {return browser.$('.close-dialog');}
+
+    get errorMessage(){return browser.$('.error')};
+
+    async navigateToLoginPage() {
+        await this.navigateTo('http://demo.owasp-juice.shop/#/login');
+        return this.waitForElement(this.welcomeDialog);
+    }
+
+    clickLoginButton() {
+        return this.clickOnElement(this.loginButton);
+    }
+
+    closeWelcomeDialog() {
+        return this.clickOnElement(this.dismissButton);
+    }
+
+    async fillLoginForm(data = {}) {
```

Comment:
Interesting solution but I would not recommend to use it on other projects. There are several reasons.
1. You will get many eslint errors
2. If you change selector name, you won't be able to track this error using built-in linters
3. 2nd point is actual for the data object 
Please take it into account

Reply:
1. We can configure eslint rule to our requirement. 
2. We have the same problem now, isn't it? 

In this case i agree that we can use default approach but i think that this approach make sense to use to another project even in our project too. 
Imagine you have page where you need to fill 20 or more fields then method which fill data in the field will be on 20 stings. 
So this approach will help to reduce 20 lines of codes to 4 -5 lines.

Reply:
1. It would be difficult to implement this because the eslint features are limited. At least, as far as I know
2.  Actually, no. Currently, unused variable will be highlighted and you notice that something went wrong. One more thing, you can easily find references for a selected variable. When you don't use it explicitly, you cannot do the same.
Regarding number of fields, so far I have not seen a form where more than 10 fields were required. It has a big impact on user experience and because of this, not all fields are required in this case. With non-required fields you can work in the separate spec blocks. The last one argument, if you have requirements to the naming, it will cause additional troubles. Consider below example:
```
get addBookModalTitleField() { return $('input[name=\'title\']'); }
get addBookModalAuthorField() { return $('input[name=\'author\']'); }
get addBookModalYearOfPublishingField() { return $('input[name=\'year\']'); }
get addBookModalNumberOfPagesField() { return $('input[name=\'pages\']'); }
get addBookModalLanguageField() { return $('input[name=\'language\']'); }

fillAddNewBookForm(data) {
    for (const key of Object.keys(data)) {
         await this.changeElementText(this[key], data[key]);
     }
     return this;
    }
}
```
Call method from the test:
```
it('Should add new book', async() => {
  const bookInfo = {addBookModalTitleField: 'Harry Potter', addBookModalAuthorField: 'J. K. Rowling', addBookModalYearOfPublishingField: '1999', addBookModalNumberOfPagesField: '545', addBookModalLanguageField: 'en'};
   await BookPage.fillAddNewBookForm(bookInfo);
   expect(BookPage.getBooksListLength()).to.eventually.equal(1);
});
```
VS
```
it('Should add new book', async() => {
  const bookInfo = {title: 'Harry Potter', author: 'J. K. Rowling', yearOfPublishing: '1999', numberOfPages: '545', language: 'en'};
   await BookPage.fillAddNewBookForm(bookInfo);
   expect(BookPage.getBooksListLength()).to.eventually.equal(1);
});
```
Now imagine that you should create an object as in the first example for 20 fields (what about KISS principle??) )))
At the end of the day, it is only my opinion and if the proposed approach is better for you, fell free to use it.

Reply:
For bookInfo object will be created dataGenerator method where you create the object with 20 fields, i think you understand this and you won't see that data in test. My goal was to try another approach to fill form and i tried to do it. I think that in some cases it good approach to use it. 
So let's stop arguing =)